File: src/shared/e2ee/cipher.ts
/*
 * taskwraith-e2ee-v1 — AES-256-GCM seal/open with directional nonces + AAD.
 */

import { createCipheriv, createDecipheriv } from 'crypto'
import { NONCE_PREFIX, type Direction } from './protocol'

const NONCE_LEN = 12
const TAG_LEN = 16

/** 12-byte nonce = 4B big-endian direction prefix ‖ 8B big-endian seq. */
export function buildNonce(direction: Direction, seq: number): Buffer {
  const nonce = Buffer.alloc(NONCE_LEN)
  nonce.writeUInt32BE(NONCE_PREFIX[direction], 0)
  nonce.writeBigUInt64BE(BigInt(seq), 4)
  return nonce
}

/** AAD binds ciphertext to its session + position (defeats cross-session/replay splices). */
export function buildAad(sessionId: string, seq: number): Buffer {
  return Buffer.from(`${sessionId}:${seq}`, 'utf8')
}

export interface SealedFrame {
  nonce: Buffer
  ct: Buffer
  tag: Buffer
}

export function seal(
  key: Buffer,
  direction: Direction,
  sessionId: string,
  seq: number,
  plaintext: Buffer
): SealedFrame {
  const nonce = buildNonce(direction, seq)
  const cipher = createCipheriv('aes-256-gcm', key, nonce)
  cipher.setAAD(buildAad(sessionId, seq))
  const ct = Buffer.concat([cipher.update(plaintext), cipher.final()])
  return { nonce, ct, tag: cipher.getAuthTag() }
}

/**
 * Decrypt + authenticate. Recomputes the expected nonce from (direction, seq)
 * and rejects a mismatched wire nonce before doing anything else; the GCM tag
 * check (bound to the AAD) then rejects any tamper. Throws on any failure.
 */
export function open(
  key: Buffer,
  direction: Direction,
  sessionId: string,
  seq: number,
  frame: { nonce: Buffer; ct: Buffer; tag: Buffer }
): Buffer {
  const expected = buildNonce(direction, seq)
  if (!frame.nonce.equals(expected)) throw new Error('nonce mismatch')
  if (frame.tag.length !== TAG_LEN) throw new Error('bad tag length')
  const decipher = createDecipheriv('aes-256-gcm', key, expected)
  decipher.setAAD(buildAad(sessionId, seq))
  decipher.setAuthTag(frame.tag)
  return Buffer.concat([decipher.update(frame.ct), decipher.final()])
}

File: ios/TaskWraithKit/Sources/TaskWraithKit/Cipher.swift
// taskwraith-e2ee-v1 — AES-256-GCM seal/open (Swift port of
// src/shared/e2ee/cipher.ts). 12-byte nonce = 4B BE direction prefix ‖ 8B BE
// seq; AAD = "sessionId:seq". CryptoKit's default GCM tag is 128-bit, matching
// Node's getAuthTag().

import Foundation
import CryptoKit

public struct SealedFrame: Sendable {
    public var nonce: Data
    public var ct: Data
    public var tag: Data
}

public enum TWCipher {
    public enum CipherError: Error { case nonceMismatch, badTagLength, openFailed }

    public static func buildNonce(_ direction: Direction, _ seq: Int) -> Data {
        var nonce = Data(count: 12)
        let prefix = direction.noncePrefix.bigEndian
        withUnsafeBytes(of: prefix) { nonce.replaceSubrange(0..<4, with: $0) }
        let seqBE = UInt64(seq).bigEndian
        withUnsafeBytes(of: seqBE) { nonce.replaceSubrange(4..<12, with: $0) }
        return nonce
    }

    public static func buildAad(_ sessionId: String, _ seq: Int) -> Data {
        Data("\(sessionId):\(seq)".utf8)
    }

    public static func seal(
        key: SymmetricKey, direction: Direction, sessionId: String, seq: Int, plaintext: Data
    ) throws -> SealedFrame {
        let nonce = buildNonce(direction, seq)
        let sealedBox = try AES.GCM.seal(
            plaintext, using: key, nonce: AES.GCM.Nonce(data: nonce),
            authenticating: buildAad(sessionId, seq))
        return SealedFrame(nonce: nonce, ct: sealedBox.ciphertext, tag: sealedBox.tag)
    }

    public static func open(
        key: SymmetricKey, direction: Direction, sessionId: String, seq: Int, frame: SealedFrame
    ) throws -> Data {
        let expected = buildNonce(direction, seq)
        guard frame.nonce == expected else { throw CipherError.nonceMismatch }
        guard frame.tag.count == 16 else { throw CipherError.badTagLength }
        let box = try AES.GCM.SealedBox(
            nonce: AES.GCM.Nonce(data: expected), ciphertext: frame.ct, tag: frame.tag)
        return try AES.GCM.open(box, using: key, authenticating: buildAad(sessionId, seq))
    }
}
